### ISSUE TITLE
Bazel 7: `--sandbox_add_mount_pair` under `/tmp` fails

### ISSUE THREAD

**Issue by fmeum** [bazelbuild/bazel#20527]
### Description of the bug:

With Bazel 7, but neither Bazel 6.4.0 nor `--noincompatible_sandbox_hermetic_tmp`, builds with `--sandbox_add_mount_pair` referencing a path under `/tmp` fail since `/tmp` has been remounted before the manually specified mount pair is applied.

### Which category does this issue belong to?

Local Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

```shell
touch WORKSPACE

cat > .bazelrc <<'EOF'
build --sandbox_add_mount_pair=/tmp/some/path:/etc
EOF

cat > BUILD <<'EOF'
genrule(
    name = "gen",
    outs = ["data.txt"],
    cmd = "ls /etc > $@",
)
EOF
```

Then:
```
$ mkdir -p /tmp/some/path
$ bazel clean --expunge && bazel shutdown && bazel build //:gen
...
src/main/tools/linux-sandbox-pid1.cc:305: "mount(/tmp/some/path, /etc, nullptr, MS_BIND | MS_REC, nullptr)": No such file or directory
Target //:gen failed to build
...
# bazel clean --expunge && bazel shutdown && bazel build //:gen --noincompatible_sandbox_hermetic_tmp
...
INFO: Build completed successfully, 2 total actions
```

### Which operating system are you running Bazel on?

Linux

### What is the output of `bazel info release`?

7.0.0

**Comment by fmeum** [comment]
CC @lberki 

A simple workaround would be to just disable the hermetic `/tmp` in this case, similar to what we do for tmpfs paths. For example:
```java
diff --git a/src/main/java/com/google/devtools/build/lib/sandbox/LinuxSandboxedSpawnRunner.java b/src/main/java/com/google/devtools/build/lib/sandbox/LinuxSandboxedSpawnRunner.java
index 3f6e49c72c..1e9b58ae00 100644
@@ -206,17 +207,21 @@ final class LinuxSandboxedSpawnRunner extends AbstractSandboxSpawnRunner {
       return false;
     }
 
-    Optional<PathFragment> tmpfsPathUnderTmp =
-        getSandboxOptions().sandboxTmpfsPath.stream()
+    Optional<PathFragment> mountUnderTmp =
+        Stream.concat(
+                getSandboxOptions().sandboxTmpfsPath.stream(),
+                getSandboxOptions().sandboxAdditionalMounts.stream()
+                    .map(Map.Entry::getKey)
+                    .map(PathFragment::create))
             .filter(path -> path.startsWith(SLASH_TMP))
             .findFirst();
-    if (tmpfsPathUnderTmp.isPresent()) {
+    if (mountUnderTmp.isPresent()) {
       if (warnedAboutNonHermeticTmp.compareAndSet(false, true)) {
         reporter.handle(
             Event.warn(
                 String.format(
-                    "Falling back to non-hermetic '/tmp' in sandbox due to '%s' being a tmpfs path",
-                    tmpfsPathUnderTmp.get())));
+                    "Falling back to non-hermetic '/tmp' in sandbox due to '%s' being a tmpfs path or mount source.",
+                    mountUnderTmp.get())));
       }
 
       return false;
```

**Comment by fmeum** [comment]
Flipping source and target in my example also results in a failure, not sure why I couldn't reproduce this earlier. I have edited the issue description accordingly.

**Comment by fmeum** [comment]
@bazel-io flag

**Comment by fmeum** [comment]
@bazel-io fork 7.0.1

**Comment by lberki** [comment]
https://github.com/bazelbuild/bazel/pull/20603 is a draft fix for this. I'm not sure if it works, but I'll ask @tjgq .

**Comment by lberki** [comment]
I have a fix for this one. It's a one-line fix, modulo the test.

**Comment by fmeum** [comment]
@lberki Feel free to reuse the integration tests from https://github.com/bazelbuild/bazel/pull/20583.

**Comment by lberki** [comment]
Sorry, too late, I already wrote my own :) (very similar to yours, though)

**Comment by lberki** [comment]
Mount *targets*? I seem to remember discussing this with you at some point in time, and IIRC the consensus was that mount targets under `/tmp` aren't a good idea: the whole point of `--incompatible_sandbox_hermetic_tmp` is that `/tmp` is pristine and mount points under it seem to compromise that.

**Comment by fmeum** [comment]
I would say that's still entirely true, but with the flag flip, the user is no longer the one who specified both `--incompatible_sandbox_hermetic_tmp` *and* the mount pair. I think that it would be entirely reasonable to just automatically disable the hermetic `/tmp` in that case, but not doing so results in a confusing error.

**Comment by lberki** [comment]
WDYT about emitting a non-confusing error?

I agree that the current situation isn't fantastic and technically speaking, allowing mount targets under `/tmp` is not impossible, but I'd rather not allow that, at least on the first attempt.

**Comment by iancha1992** [comment]
@bazel-io fork 7.1.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.0.1 RC2](https://releases.bazel.build/7.0.1/rc2/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!
